odl/trafos/wavelet.py

```diff
@@ -834,14 +957,511 @@ def adjoint(self):
+                             'got {}.'.format(max_level, self.nscales))
+            #raiseETC ('Setting {} to {}' .format(self.nscales, max_level))
+            #nscales = max_level
+
+        filterlength = int(wbasis[9])
+        if filterlength not in (1, 3, 5, 7, 9):
+            raise NotImplementedError('Filterlength {} not 1, 3, 5, 7 or 9'
+                                      ''.format(filterlength))
+
+        dom_size = ran.size
+        dom = ran.dspace_type(dom_size, dtype=ran.dtype)
+        super().__init__(dom, ran, linear=True)
+
+    def _call(self, coeff):
+        """Discrete inverse wavelet transform with biorthogonal wavelets.
+
+        Parameters
+        ----------
+        coeff : `DiscreteLpVector`
+
+        Returns
+        -------
+        arr : `DiscreteLpVector`
+        """
+        filterlength = int(self.wbasis[9])
+        if len(self.range.grid.shape) == 1:
+            nx = self.range.grid.shape[0]
+            coeff_in = coeff.copy()
+            coeff_in = coeff_in.asarray()
+            x = np.asarray(self.range.element())
+            bwt.invwavelet_transform1D(coeff_in.ctypes.data, nx,
+                                       filterlength,
+                                       self.nscales, x.ctypes.data)
+            return self.range.element(x)
+
+        elif len(self.range.grid.shape) == 2:
+            (nx, ny) = (self.range.grid.shape[0], self.range.grid.shape[1])
+            coeff_in = coeff.copy()
+            coeff_in = coeff_in.asarray()
+            x = np.asarray(self.range.element())
+            bwt.invwavelet_transform2D(coeff_in.ctypes.data, nx, ny,
+                                       filterlength,
+                                       self.nscales, x.ctypes.data)
+            return self.range.element(x)
+
+        elif len(self.range.grid.shape) == 3:
+            (nx, ny, nz) = (self.range.grid.shape[0], self.range.grid.shape[1],
+                            self.range.grid.shape[2])
+            coeff_in = coeff.copy()
+            coeff_in = coeff_in.asarray()
+            x = np.asarray(self.range.element())
+            bwt.invwavelet_transform3D(coeff_in.ctypes.data, nx, ny, nz,
+                                       filterlength,
+                                       self.nscales, x.ctypes.data)
+            return self.range.element(x)
+
+    @property
+    def adjoint(self):
+        """The biorthogonal wavelet transform."""
+        return InverseAdjBiorthWaveletTransform(dom=self.range,
+                                                nscales=self.nscales,
+                                                wbasis=self.wbasis)
+
+    @property
+    def inverse(self):
+        """The inverse of the adjoint of biorthogonal wavelet transform."""
+        return BiorthWaveletTransform(dom=self.range, nscales=self.nscales,
+                                      wbasis=self.wbasis)
+
+
+class InverseAdjBiorthWaveletTransform(Operator):
```

Comment:
These four classes are all terribly similar, is there no chance to change them to one class?